posts/inside-rust/2021-05-15-What-the-error-handling-project-group-is-working-towards.md

```diff
@@ -0,0 +1,235 @@
+
+Consider this example:
+
+```rust
+use std::fmt;
+
+// We have a program that loads a config and expects that
+// loading the config will always succeed.
+fn main() {
+    let _config = load_config()
+        .expect("config is always valid and exists");
+}
+
+// We have a dummy implementation of load_config which
+// always errors, since we're just focusing on diagnostics
+// here.
+fn load_config() -> Result<(), Error> {
+    Err(Error(SourceError))
+}
+
+// And we have an error type that just prints "invalid
+// config" and has a source error which just prints "config
+// file does not exist"
+#[derive(Debug)]
+struct Error(SourceError);
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("invalid config")
+    }
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        Some(&self.0)
+    }
+}
+
+#[derive(Debug)]
+struct SourceError;
+
+impl fmt::Display for SourceError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("config file does not exist")
+    }
+}
+
+impl std::error::Error for SourceError {}
+```
+
+When we run this we would like to see output somewhat like this:
+
+```
+$ cargo run
+thread 'main' panicked at 'config is always valid and exists', src/main.rs:4:33
+
+Error:
+    0: invalid config
+    1: config file does not exist
+
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+```
+
+In this error message we can see that we exited because of a panic. We can see what invariant we violated that was supposed to prevent this panic from happening. We can see the location where the panic was produced. And we can see the error message of each error in the chain of errors accessible via `source`.
+
+That's what we would like, at least in the version of rust that the error handling project group wants to see, but what we actually get is this...
+
+```
+$ cargo run
+thread 'main' panicked at 'config is always valid and exists: Error(SourceError)', main.rs:4:33
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+```
+
+Now, I definitely don't think this is what we want as a default when promoting recoverable errors to non-recoverable errors! `unwrap` and `expect` work by stringifying the error variant using it's `Debug` impl, but this is often the wrong operation for types that implement the `Error` trait. By converting the `Error` to a `String` we lose access to the pieces of context we carefully split up via the `Error` trait, and in all likelihood the `derive(Debug)` output of our error types won't even include the error messages in our `Display` impls. Rust's panic infrastructure doesn't provide a method for converting an `Error` type into a panic, it only supports converting `Debug` types into panics, and we feel that this is a major issue.
```

Comment:
a newline here would make this more expressive

Reply:
I'm not sure exactly where you were referring to but I took a guess, let me know if it's better.

Reply:
much better :)